Create Plainbox Provider plugin

jocave opened the pull request:
lp: #1552369

It would be useful if this plugin could be released on to the 1.x series as well.

snappy-m-o commented:
Can one of the admins verify this patch?

snappy-m-o commented:
Can one of the admins verify this patch?

zyga commented on snapcraft/plugins/plainbox_provider.py:
```diff
@@ -0,0 +1,57 @@
+deb http://${security}.ubuntu.com/${suffix} ${release}-security main restricted
+deb http://${security}.ubuntu.com/${suffix} ${release}-security universe
+deb http://${security}.ubuntu.com/${suffix} ${release}-security multiverse
+deb http://ppa.launchpad.net/checkbox-dev/ppa/ubuntu ${release} main
```
I'd use stable PPA here really or make this customizable with an explicit knob.

zyga commented on integration_tests/snaps/simple-plainbox-provider/2016.com.example:simple/manage.py:
```diff
@@ -0,0 +1,10 @@
+
+
+setup(
+    name='2016.com.example:simple',
```
I'd use ``plainbox-provider-example`` here with an explicit ``namespace="2016.com.example"``

jocave commented:
I've updated the code to make it possible to disable the Checkbox Development PPA. Also, switched to using an explicit namespace (kept the provider name "Simple" though").

elopio commented:
OK to test

elopio commented:
retest this please

jocave commented:
retest this please

elopio commented:
retest this please

elopio commented:
add to whitelist

elopio commented:
retest this please

elopio commented on snapcraft/plugins/plainbox_provider.py:
```diff
@@ -0,0 +1,76 @@
+        super().__init__(name, options)
+        self.build_packages.extend(['python3-plainbox', 'intltool'])
+        if self.options.checkbox_dev_ppa:
+            self._PLUGIN_STAGE_SOURCES = self._DEV_PPA_PLUGIN_STAGE_SOURCES
```
this _PLUGIN_STAGE_SOURCES is not a constant, so it shouldn't be all capital letters.
_plugin_stage_sources.

elopio replied:
sorry, not from your branch. After you merge I'll propose something to change it.

For you, it seems safer to overwrite the PLUGIN_STAGE_SOURCES @property, instead of dealing with the internal attribute.

sergiusens replied:
Yes, don't take advantage of internal implementation details which are subject to change :wink:

elopio commented:
:+1: from me.
Can you please remove the icon? It's not needed, and I think it's better to keep the test snaps with only the required bits.

sergiusens commented on integration_tests/snaps/simple-plainbox-provider/snapcraft.yaml:
```diff
@@ -0,0 +1,12 @@
+description: |
+    Create a snap of a very simple plainbox that could
+    then be used by a checkbox application
+icon: icon.png
```
remove this entry please

sergiusens commented:
`git rm integration_tests/snaps/simple-plainbox-provider/icon.png`

jocave commented:
Pushed commit with icon file removed from simple-plainbox-provider integration test.

sergiusens commented:
`git mv integration_tests/test_plainbox_provider_plugin.py integration_tests/test_plugin_plainbox_provider.py`

sergiusens commented on snapcraft/plugins/plainbox_provider.py:
```diff
@@ -0,0 +1,76 @@
+        schema['properties']['checkbox-dev-ppa'] = {
+            'type': 'boolean',
+            'default': 'true',
+        }
```
since you aren't overriding `properties` it seems you are making use of the common `source` options, can you add that piece of information to the doc string?

sergiusens commented on snapcraft/plugins/plainbox_provider.py:
```diff
@@ -0,0 +1,76 @@
+
+    def __init__(self, name, options):
+        super().__init__(name, options)
+        self.build_packages.extend(['python3-plainbox', 'intltool'])
```
`build-packages` are installed onto the host from the hosts defined `sources.list[.d]`; in the documentation above you mention that plainbox features are not completely in the archive. How does adding `python3-plainbox` here play into this?

sergiusens commented:
Thanks for this, I have a bunch of open questions in there though.

Also, does this rely on python provided by the os snap? It is fine from my PoV, just want to make sure.

sergiusens commented:
Given the previous comment I would like to see an example and an example_test (which would run on a snappy system)

elopio commented:
The example test would be so nice, indeed.
@jocave take a look at the examples_tests dir. Give me a shout if you need a hand with that one.

jocave commented:
Thank you for the reviews, I will put an example together and try and address the questions ASAP.

jocave commented:
I have added integration tests to cover a set of what I think are expected use cases for this plugin. One of these fails with a python import error for module 'plainbox'. This is provided by the package python3-plainbox which is explicitly added to the build-package list by the plugin.

I would like to know how to guarantee that 'build-package' works as advertised.

sergiusens commented:
@jocave build-packages will just land on the host and be used to build the system, there is additional magic to crawl everything ELF and bring in any .so it needs not provided in the snap/part area.

stage-packages however, which is what you use in your examples are a bit tricky, specially when it comes to python as you will need to define a `PYTHONPATH` for this.

Given this still needs a few twists and turns I am going to close it for now. Feel free to reopen when ready.
